### PR TITLE
drivers: display: max7219: add driver

### PR DESCRIPTION
**Pull request by yanagiis**
This is a continuation of PR #47416.

This PR rewrites the MAX7219 as a display driver.

### CODE REVIEW

File: samples/drivers/display/boards/stm32_min_dev_black.conf [hunk 1]
```diff
@@ -0,0 +1,2 @@
+CONFIG_HEAP_MEM_POOL_SIZE=1024
```
Comment (jfischer-no) [inline samples/drivers/display/boards/stm32_min_dev_black.conf:1]: NACK for board specific overlays in this case. Please take a look at the shield support, you can add a generic MAX7219 shield like  https://github.com/zephyrproject-rtos/zephyr/tree/main/boards/shields/ssd1306

**Review by jfischer-no — CHANGES_REQUESTED** [review]

File: samples/drivers/display/boards/stm32_min_dev_black.conf (hunk 1 above)
Comment (yanagiis) [reply samples/drivers/display/boards/stm32_min_dev_black.conf]: OK. I will add MAX7219 into `boards/shields`.
Thanks.

File: drivers/display/display_max7219.c [hunk 2]
```diff
@@ -0,0 +1,385 @@
...
+
+	for (int i = 0; i < dev_config->num_cascading; i++) {
+		dev_data->tx_buf[i * 2] = addr;
+		dev_data->tx_buf[i * 2 + 1] = value;
+	}
+
+	return spi_write_dt(&dev_config->spi, &tx_bufs);
+}
+
+static int max7219_transmit_one(const struct device *dev, uint8_t max7219_idx, uint8_t addr,
```
Comment (jfischer-no) [inline drivers/display/display_max7219.c:92]:
```diff
diff --git a/drivers/display/display_max7219.c b/drivers/display/display_max7219.c
index e79c1cb6bee..5422072b508 100644
--- a/drivers/display/display_max7219.c
+++ b/drivers/display/display_max7219.c
@@ -89,8 +89,8 @@ static int max7219_transmit_all(const struct device *dev, uint8_t addr, uint8_t
        return spi_write_dt(&dev_config->spi, &tx_bufs);
 }
 
-static int max7219_transmit_one(const struct device *dev, uint8_t max7219_idx, uint8_t addr,
-                               uint8_t value)
+static int max7219_transmit_one(const struct device *dev, const uint8_t max7219_idx,
+                               const uint8_t addr, const uint8_t value)
 {
        const struct max7219_config *dev_config = dev->config;
        struct max7219_data *dev_data = dev->data;
```

File: drivers/display/display_max7219.c [hunk 3]
```diff
@@ -0,0 +1,385 @@
...
+		}
+
+		dev_data->tx_buf[i * 2] = addr;
+		dev_data->tx_buf[i * 2 + 1] = value;
+	}
+
+	return spi_write_dt(&dev_config->spi, &tx_bufs);
+}
+
+static inline uint8_t next_pixel(uint8_t *mask, uint8_t *data, const uint8_t **byte_buf)
```
Comment (jfischer-no) [inline drivers/display/display_max7219.c:121]: `s/byte_buf/buf/ `?

File: drivers/display/display_max7219.c [hunk 4]
```diff
@@ -0,0 +1,385 @@
...
+	static const struct max7219_config max7219_config_##n = {                                  \
+		.spi = SPI_DT_SPEC_INST_GET(n, SPI_OP_MODE_MASTER | SPI_WORD_SET(8U), 0U),         \
+		.num_cascading = DT_INST_PROP(n, num_cascading),                                   \
+		.intensity = DT_INST_PROP(n, intensity),                                           \
+		.scan_limit = DT_INST_PROP(n, scan_limit),                                         \
+	};                                                                                         \
+	DEVICE_DT_INST_DEFINE(n, max7219_init, NULL, &max7219_data_##n, &max7219_config_##n,       \
+			      POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY, &max7219_api);
+
+DT_INST_FOREACH_STATUS_OKAY(DISPLAY_MAX7219_INIT)
```
Comment (jfischer-no) [inline drivers/display/display_max7219.c:385]: please indent consistently with tabs

```diff
diff --git a/drivers/display/display_max7219.c b/drivers/display/display_max7219.c
index e79c1cb6bee..00e4dae1c55 100644
--- a/drivers/display/display_max7219.c
+++ b/drivers/display/display_max7219.c
@@ -365,21 +365,25 @@ static int max7219_init(const struct device *dev)
        return 0;
 }
 
-#define DISPLAY_MAX7219_INIT(n)                                                                    \
-       static uint8_t max7219_digit_data_##n[DT_INST_PROP(n, num_cascading) *                     \
-                                             MAX7219_DIGITS_PER_DEVICE];                          \
-       static uint8_t max7219_tx_buf##n[DT_INST_PROP(n, num_cascading) * 2];                      \
-       static struct max7219_data max7219_data_##n = {                                            \
-               .digit_buf = max7219_digit_data_##n,                                               \
-               .tx_buf = max7219_tx_buf##n,                                                       \
-       };                                                                                         \
-       static const struct max7219_config max7219_config_##n = {                                  \
-               .spi = SPI_DT_SPEC_INST_GET(n, SPI_OP_MODE_MASTER | SPI_WORD_SET(8U), 0U),         \
-               .num_cascading = DT_INST_PROP(n, num_cascading),                                   \
-               .intensity = DT_INST_PROP(n, intensity),                                           \
-               .scan_limit = DT_INST_PROP(n, scan_limit),                                         \
-       };                                                                                         \
-       DEVICE_DT_INST_DEFINE(n, max7219_init, NULL, &max7219_data_##n, &max7219_config_##n,       \
-                             POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY, &max7219_api);
+#define DISPLAY_MAX7219_INIT(n)                                                        \
+       static uint8_t max7219_digit_data_##n[DT_INST_PROP(n, num_cascading) *  \
+                                             MAX7219_DIGITS_PER_DEVICE];       \
+       static uint8_t max7219_tx_buf##n[DT_INST_PROP(n, num_cascading) * 2];   \
+       static struct max7219_data max7219_data_##n = {                         \
+               .digit_buf = max7219_digit_data_##n,                            \
+               .tx_buf = max7219_tx_buf##n,                                    \
+       };                                                                      \
+       static const struct max7219_config max7219_config_##n = {               \
+               .spi = SPI_DT_SPEC_INST_GET(n, SPI_OP_MODE_MASTER |             \
+                                           SPI_WORD_SET(8U), 0U),              \
+               .num_cascading = DT_INST_PROP(n, num_cascading),                \
+               .intensity = DT_INST_PROP(n, intensity),                        \
+               .scan_limit = DT_INST_PROP(n, scan_limit),                      \
+       };                                                                      \
+                                                                               \
+       DEVICE_DT_INST_DEFINE(n, max7219_init, NULL,                            \
+                             &max7219_data_##n, &max7219_config_##n,           \
+                             POST_KERNEL, CONFIG_DISPLAY_INIT_PRIORITY,        \
+                             &max7219_api);
 
 DT_INST_FOREACH_STATUS_OKAY(DISPLAY_MAX7219_INIT)
```

File: drivers/display/display_max7219.c [hunk 5]
```diff
@@ -0,0 +1,385 @@
...
+ * Two MAX7219s (with cascading) have 8x16 pixels.
+ * So on and so forth.
+ *
+ * Datasheet: https://datasheets.maximintegrated.com/en/ds/MAX7219-MAX7221.pdf
+ *
+ * Limitations:
+ *  1. This driver only implements no-decode mode.
+ */
+
+#include <stddef.h>
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply drivers/display/display_max7219.c:25]: MAX7219 has a register called decode-mode. 
The decode mode is useful when MAX7219 is connecting to 7-segment LEDs.
This driver set this register as no-decode mode now.

File: drivers/display/display_max7219.c [hunk 6]
```diff
@@ -0,0 +1,385 @@
...
+
+#define MAX7219_REG_SCAN_LIMIT 0x0B
+
+#define MAX7219_REG_SHUTDOWN	    0x0C
+#define MAX7219_SHUTDOWN_MODE	    0x00
+#define MAX7219_LEAVE_SHUTDOWN_MODE 0x01
+
+#define MAX7219_REG_DISPLAY_TEST	0x0F
+#define MAX7219_LEAVE_DISPLAY_TEST_MODE 0x00
+#define MAX7219_DISPLAY_TEST_MODE	0x01
```
[the remark this answers is not in the document]
Comment (yanagiis) [reply drivers/display/display_max7219.c:55]: I have used the `clang-format` format this file. Should I add `clang-format off` for this block?

File: drivers/display/display_max7219.c (hunk 3 above)
Comment (yanagiis) [reply drivers/display/display_max7219.c:121]: Yes. I will change it.

File: drivers/display/display_max7219.c (hunk 4 above)
Comment (yanagiis) [reply drivers/display/display_max7219.c:385]: Thanks. I have used the clang-format format this file. Should I add clang-format off for this block? or follow clang-format config in zephyr?

Comment (jfischer-no) [reply drivers/display/display_max7219.c:388]: > Thanks. I have used the `clang-format` format this file. Do you also use the `clang-format`?

No, it is just Vim and me.

**Review by jfischer-no — APPROVED** [review]

**Review by carlescufi — APPROVED** [review]

**Pull request merged**